File: backend/app/documents.py
"""Document-type registry and prompt/schema builders (PL-6).

The registry itself is the shared ``documents.json`` (single source of truth,
also consumed by the frontend). This module loads it and derives, per document:

  * the selection catalogue used when the AI is figuring out *which* document
    the user wants;
  * a flat list of leaf fields (``party`` terms expand into four sub-fields);
  * a dynamic Pydantic model for structured extraction of a generic document's
    fields.
"""

from __future__ import annotations

import json
from functools import lru_cache
from typing import Optional

from pydantic import BaseModel, create_model

from .config import DOCUMENTS_PATH

# Sub-fields a ``party`` key expands into — must match the frontend's
# PARTY_SUBFIELDS in lib/documents.ts.
PARTY_SUBFIELDS: list[tuple[str, str]] = [
    ("company", "Company"),
    ("name", "Print Name"),
    ("title", "Title"),
    ("notice", "Notice Address"),
]


@lru_cache(maxsize=1)
def load_documents() -> list[dict]:
    """All registered documents (cached). Reads the shared documents.json."""
    raw = json.loads(DOCUMENTS_PATH.read_text(encoding="utf-8"))
    return raw["documents"]


def get_document(doc_id: str) -> Optional[dict]:
    return next((d for d in load_documents() if d["id"] == doc_id), None)


def selection_catalogue() -> str:
    """A bulleted ``id — name: description`` list for the selection prompt."""
    return "\n".join(
        f"- {d['id']} — {d['name']}: {d['description']}" for d in load_documents()
    )


def leaf_fields(spec: dict) -> list[dict]:
    """Flatten a generic document's keyTerms into leaf fields.

    Each entry: {key, label, hint?, type, options?}. A ``party`` term expands
    into four leaf fields (``<key>_company`` etc.).
    """
    leaves: list[dict] = []
    for term in spec.get("keyTerms", []):
        if term["type"] == "party":
            for suffix, sub_label in PARTY_SUBFIELDS:
                leaves.append(
                    {
                        "key": f"{term['key']}_{suffix}",
                        "label": f"{term['label']} — {sub_label}",
                        "type": "text",
                    }
                )
        else:
            leaves.append(
                {
                    "key": term["key"],
                    "label": term["label"],
                    "hint": term.get("hint"),
                    "type": term["type"],
                    "options": term.get("options"),
                }
            )
    return leaves


@lru_cache(maxsize=None)
def generic_fields_model(doc_id: str) -> type[BaseModel]:
    """A dynamic Pydantic model: one Optional[str] per leaf field of ``doc_id``.

    Used as the structured-output schema so the LLM returns only this document's
    known keys (never free-form), which we can merge straight into the preview.
    """
    spec = get_document(doc_id)
    if spec is None:
        raise KeyError(doc_id)
    field_defs = {
        leaf["key"]: (Optional[str], None) for leaf in leaf_fields(spec)
    }
    safe = doc_id.replace("-", "_").title().replace("_", "")
    return create_model(f"{safe}Fields", **field_defs)  # type: ignore[call-overload]


def generic_fields_prompt(spec: dict) -> str:
    """Human-readable field list injected into the generic drafting prompt."""
    lines: list[str] = []
    for leaf in leaf_fields(spec):
        bits = [f"- {leaf['key']}: {leaf['label']}"]
        if leaf.get("hint"):
            bits.append(f"({leaf['hint']})")
        if leaf["type"] == "date":
            bits.append('— ISO "yyyy-mm-dd"')
        if leaf.get("options"):
            bits.append("— one of: " + ", ".join(leaf["options"]))
        lines.append(" ".join(bits))
    return "\n".join(lines)

File: frontend/lib/documents.ts
/**
 * Document-type registry (PL-6).
 *
 * Loads the shared `documents.json` (the single source of truth, also read by
 * the backend) and exposes typed helpers. `nda` documents use the bespoke
 * Mutual NDA pipeline; `generic` documents are driven entirely by their
 * `keyTerms` — collected over chat and rendered by the generic preview / PDF.
 */

import documentsData from "./documents.json";

export type FieldType =
  | "text"
  | "textarea"
  | "date"
  | "number"
  | "select"
  | "party";

export interface KeyTerm {
  key: string;
  label: string;
  type: FieldType;
  hint?: string;
  options?: string[];
}

export type DocumentKind = "nda" | "generic";

export interface DocumentSpec {
  id: string;
  name: string;
  kind: DocumentKind;
  description: string;
  /** Present for `generic` documents; absent for the bespoke NDA. */
  keyTerms?: KeyTerm[];
}

/** Flat string map of a generic document's collected values. */
export type GenericDocData = Record<string, string>;

/** Sub-fields a `party` key expands into (suffix appended as `<key>_<suffix>`). */
export const PARTY_SUBFIELDS: { suffix: string; label: string }[] = [
  { suffix: "company", label: "Company" },
  { suffix: "name", label: "Print Name" },
  { suffix: "title", label: "Title" },
  { suffix: "notice", label: "Notice Address" },
];

export const DOCUMENTS: DocumentSpec[] = (
  documentsData as { documents: DocumentSpec[] }
).documents;

export function getDocument(id: string): DocumentSpec | undefined {
  return DOCUMENTS.find((d) => d.id === id);
}

/**
 * The leaf data keys for a generic document, expanding every `party` field into
 * its four sub-fields. Order follows the keyTerms, sub-fields inner.
 */
export function leafKeys(spec: DocumentSpec): string[] {
  const keys: string[] = [];
  for (const t of spec.keyTerms ?? []) {
    if (t.type === "party") {
      for (const sub of PARTY_SUBFIELDS) keys.push(`${t.key}_${sub.suffix}`);
    } else {
      keys.push(t.key);
    }
  }
  return keys;
}

/** Empty starting values for a generic document (every leaf key blank). */
export function emptyGenericData(spec: DocumentSpec): GenericDocData {
  const data: GenericDocData = {};
  for (const k of leafKeys(spec)) data[k] = "";
  return data;
}

/**
 * Merge AI-extracted updates into generic data. Only keys that belong to the
 * document and carry a non-empty value override existing ones, so the model
 * can never wipe a field by returning a blank.
 */
export function mergeGenericFields(
  spec: DocumentSpec,
  data: GenericDocData,
  updates: GenericDocData | null | undefined,
): GenericDocData {
  if (!updates) return data;
  const allowed = new Set(leafKeys(spec));
  const next = { ...data };
  for (const [k, v] of Object.entries(updates)) {
    if (allowed.has(k) && v != null && String(v).trim() !== "") {
      next[k] = String(v);
    }
  }
  return next;
}
